lite/client/provider.go

```diff
@@ -23,119 +23,114 @@ type SignStatusClient interface {
 	if err != nil {
-		return fc, err
+		return nil, err
 	}
-	return p.seedFromCommit(commit)
-}
-
-// GetLatestCommit should return the most recent commit there is,
-// which handles queries for future heights as per the semantics
-// of GetByHeight.
-func (p *provider) GetLatestCommit() (*ctypes.ResultCommit, error) {
-	status, err := p.node.Status()
-	if err != nil {
+	if status.SyncInfo.LatestBlockHeight < minHeight {
+		err = fmt.Errorf("provider is at %v but require minHeight=%v",
+			status.SyncInfo.LatestBlockHeight, minHeight)
 		return nil, err
 	}
-	return p.node.Commit(&status.SyncInfo.LatestBlockHeight)
+	if maxHeight == 0 {
+		maxHeight = status.SyncInfo.LatestBlockHeight
+	} else if status.SyncInfo.LatestBlockHeight < maxHeight {
+		maxHeight = status.SyncInfo.LatestBlockHeight
+	}
+	return p.client.Commit(&maxHeight)
 }
 
-// CommitFromResult ...
-func CommitFromResult(result *ctypes.ResultCommit) lite.Commit {
-	return (lite.Commit)(result.SignedHeader)
+// Implements Provider.
+func (p *provider) ValidatorSet(chainID string, height int64) (valset *types.ValidatorSet, err error) {
+	return p.getValidatorSet(chainID, height)
 }
 
-func (p *provider) seedFromVals(vals *ctypes.ResultValidators) (lite.FullCommit, error) {
-	// now get the commits and build a full commit
-	commit, err := p.node.Commit(&vals.BlockHeight)
+func (p *provider) getValidatorSet(chainID string, height int64) (valset *types.ValidatorSet, err error) {
+	if chainID != p.chainID {
+		err = fmt.Errorf("expected chainID %s, got %s", p.chainID, chainID)
+		return
+	}
+	if height < 1 {
+		err = fmt.Errorf("expected height >= 1, got height %v", height)
+		return
+	}
+	heightPtr := new(int64)
+	*heightPtr = height
+	res, err := p.client.Validators(heightPtr)
 	if err != nil {
-		return lite.FullCommit{}, err
+		// TODO pass through other types of errors.
+		return nil, lerr.ErrMissingValidators(chainID, height)
 	}
-	fc := lite.NewFullCommit(
-		CommitFromResult(commit),
-		types.NewValidatorSet(vals.Validators),
-	)
-	return fc, nil
+	valset = types.NewValidatorSet(res.Validators)
+	return
 }
 
-func (p *provider) seedFromCommit(commit *ctypes.ResultCommit) (fc lite.FullCommit, err error) {
-	fc.Commit = CommitFromResult(commit)
+// This does no validation.
+func (p *provider) fillFullCommit(signedHeader types.SignedHeader) (fc lite.FullCommit, err error) {
 
-	// now get the proper validators
-	vals, err := p.node.Validators(&commit.Header.Height)
+	// Get the validators.
+	valset, err := p.getValidatorSet(signedHeader.ChainID, signedHeader.Height)
 	if err != nil {
-		return fc, err
+		return lite.FullCommit{}, err
 	}
 
-	// make sure they match the commit (as we cannot enforce height)
-	vset := types.NewValidatorSet(vals.Validators)
-	if !bytes.Equal(vset.Hash(), commit.Header.ValidatorsHash) {
-		return fc, liteErr.ErrValidatorsChanged()
+	// Get the next validators.
+	nextValset, err := p.getValidatorSet(signedHeader.ChainID, signedHeader.Height+1)
```

Comment:
NextValidatorSet is a concept in the new version. I am not sure that it's good idea to obtain it with arithmetics (height + 1) on the current validator set. Looks like exposing implementation details. Maybe we should expose NextValidatorSet in a separate call.